### PR TITLE
**new feature** in bulk resett all completion #63

### PR DESCRIPTION
**Pull request by mixedfruitjuice**
Here is my pull request for the https://github.com/danmarsden/moodle-local_recompletion/issues/63 issue :)

Please let me know if you have any feedback!

### CODE REVIEW

File: bulkresetcompletion.php [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Manually edit completion date for a course.
```
Comment (danmarsden) [inline bulkresetcompletion.php]: can you please update this comment to something like "reset completion"?

File: bulkresetcompletion.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+$context = context_course::instance($course->id);
+require_capability('local/recompletion:manage', $context);
+
+$PAGE->set_url('/local/recompletion/editcompletion.php', array('id' => $course->id));
+if (empty($users) && empty($userid)) {
+    // The first time list hack.
+    if ($post = data_submitted()) {
+        foreach ($post as $k => $v) {
+            if (preg_match('/^user(\d+)$/', $k, $m)) {
+                $users[] = $m[1];
```
Comment (danmarsden) [inline bulkresetcompletion.php]: I see this is old code copied from mine... but now that I look at that closely I don't like my code... :-)
can you please change this line to the following to cast the value from data_submitted as an int?
```
 $users[] = (int) $m[1];
```

Feel free to leave my old code on the other pages, I'll probably update that at the same time as merging this in :-)

File: editcompletion.php [hunk 3]
```diff
@@ -25,11 +25,20 @@
...
+require_once($CFG->libdir.'/completionlib.php');
+require_once($CFG->libdir.'/gradelib.php');
+require_once($CFG->dirroot . '/mod/assign/locallib.php');
+require_once($CFG->dirroot . '/mod/quiz/lib.php');
 
 $courseid = required_param('id', PARAM_INT);
 $userid   = optional_param('user', 0, PARAM_INT);
 $users   = optional_param_array('users', array(), PARAM_INT);
 
+//var_dump($users);
```
Comment (danmarsden) [inline editcompletion.php]: can you please remove this debug code?

File: editcompletion.php [hunk 4]
```diff
@@ -70,6 +79,22 @@
     $date = time();
 }
 
+//function to reset completion for $users
+if (isset($_POST['reset_completion'])) {
```
Comment (danmarsden) [inline editcompletion.php]: can you please change this one to an optional_param call? - direct access to $_POST is usually not allowed.

File: participants.php [hunk 5]
```diff
@@ -199,14 +199,22 @@
...
     echo '<input type="submit" name="submit" value="'.get_string('bulkchangedate', 'local_recompletion').'"/>';
+    echo '<input type="submit" name="reset_completion" value="'.get_string('bulkresetallcompletion', 'local_recompletion').'"/>';
+
     echo '<input type="hidden" name="id" value="' . $course->id . '" />';
     echo '</div></div>';
 }
 
 echo '</form>';
 
+
```
Comment (danmarsden) [inline participants.php]: please avoid adding extra spacing in areas that you haven't changed - can you please revert the spacing changes (including the ones below)

File: editcompletion.php (hunk 4 above)
Comment (danmarsden) [reply editcompletion.php]: it should also have a "confirm_sesskey()" check on that line as well.

**Comment by danmarsden** [comment]
thanks @mixedfruitjuice - looks like a nice feature to have - I've made some initial comments, but can you also please take a look at the coding guideline errors reported by the tests in the local/recompletion/editcompletion.php file?

**Comment by mixedfruitjuice** [comment]
@danmarsden Thanks for your feedback, I will apply your feedback and get back to you next week!

**Comment by mixedfruitjuice** [comment]
@danmarsden Sorry for the wait, due to private circumstances I wasn't able to commit this any earlier. Let me know what you think. I rewrote some code to meet the guidelines and I applied your feedback.

File: bulkresetcompletion.php [hunk 6]
```diff
@@ -0,0 +1,97 @@
+<?php
+/**
+ * Manually reset completion date for a course for a selected amount of users.
+ *
+ * @package     local_recompletion
+ * @copyright   2023 ChatGPT
```
Comment (danmarsden) [inline bulkresetcompletion.php]: I'm not sure if this was intended as a joke, but I'd prefer to have your name here or just assign it to Catalyst IT.

File: bulkresetcompletion.php (hunk 2 above)
Comment (danmarsden) [reply bulkresetcompletion.php]: doesn't look like this change was made - any chance you could squeeze that in at the same time as the other changes?

**Comment by danmarsden** [comment]
I've added a few more comments - I also note that the new code is still not passing standard Moodle codechecker stuff (shown in the failing tests above - do you have the right settings in your IDE with moodle's codechecker? - or can you please run the code through the Moodle codechecker plugin to fix it up?

Feel free to say "no" and I'll try to pick this up myself and fix it up at some point in future.

thanks!

File: bulkresetcompletion.php (hunk 6 above)
Comment (mixedfruitjuice) [reply bulkresetcompletion.php]: No, this is my mistake. I ran the code through the chatgpt plugin in phpstorm to try and see if it complies with all the Moodle coding standards and afterwards I checked it myself. But i forgot about this part where chatgpt changes the author name. Sorry about that

**Comment by gemguardian** [comment]
@danmarsden and @mixedfruitjuice thank you for working on this issue. I would greatly appriciate if this could added to the stable master in the upcoming weeks. Is that possible?

**Comment by danmarsden** [comment]
@gemguardian - I note this is still failing moodle codechecker stuff and not all the feedback I provided above has been actioned -I'll leave it in the queue here and try to come back to it one day unless @mixedfruitjuice manages to do the clean up before me.
thanks!

**Comment by PascalMasson** [comment]
Hi @danmarsden,
In consulation with @mixedfruitjuice I will we taking over this PR from him. I applied all the pending changes that I could find in the comments you placed. Please tell me if I missed anything so I can correct this.

- Pascal Masson

**Comment by danmarsden** [comment]
thanks everyone - I've tested and merged this into the 3.11 branch now.

**Pull request closed without merging**